pilot/pkg/simulation/traffic.go

```diff
@@ -0,0 +1,438 @@
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetPrefixRanges() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		ranger := cidranger.NewPCTrieRanger()
+		for _, a := range fc.GetPrefixRanges() {
+			_, cidr, err := net.ParseCIDR(fmt.Sprintf("%s/%d", a.AddressPrefix, a.GetPrefixLen().GetValue()))
+			if err != nil {
+				sim.t.Fatal(err)
+			}
+			if err := ranger.Insert(cidranger.NewBasicRangerEntry(*cidr)); err != nil {
+				sim.t.Fatal(err)
+			}
+		}
+		f, err := ranger.Contains(net.ParseIP(input.Address))
+		if err != nil {
+			sim.t.Fatal(err)
+		}
+		return f
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetServerNames() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		sni := host.Name(input.Sni)
+		for _, s := range fc.GetServerNames() {
+			if sni.SubsetOf(host.Name(s)) {
+				return true
+			}
+		}
+		return false
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetTransportProtocol() == ""
+	}, func(fc *listener.FilterChainMatch) bool {
+		switch fc.GetTransportProtocol() {
+		case xdsfilters.TLSTransportProtocol:
+			return input.Protocol == HTTPS || input.Protocol == TLS
+		case xdsfilters.RawBufferTransportProtocol:
+			return input.Protocol != HTTPS && input.Protocol != TLS
+		}
+		return false
+	})
+	chains = filter(chains, func(fc *listener.FilterChainMatch) bool {
+		return fc.GetApplicationProtocols() == nil
+	}, func(fc *listener.FilterChainMatch) bool {
+		return sets.NewSet(fc.GetApplicationProtocols()...).Contains(input.Alpn)
+	})
+	// We do not implement the "source" based filters as we do not use them
+	if len(chains) > 1 {
+		return nil, ErrMultipleFilterChain
+	}
+	if len(chains) == 0 {
+		return nil, ErrNoFilterChain
+	}
+	return chains[0], nil
+}
+
+func filter(chains []*listener.FilterChain,
+	empty func(fc *listener.FilterChainMatch) bool,
+	match func(fc *listener.FilterChainMatch) bool) []*listener.FilterChain {
+	res := []*listener.FilterChain{}
+	anySet := false
+	for _, c := range chains {
+		if !empty(c.GetFilterChainMatch()) {
+			anySet = true
+		}
+	}
+	if !anySet {
+		return chains
+	}
+	for _, c := range chains {
+		if match(c.GetFilterChainMatch()) {
+			res = append(res, c)
+		}
+	}
+	return res
+}
+
+func protocolToAlpn(s Protocol) string {
+	switch s {
+	case HTTP:
+		return "http/1.1"
+	case HTTP2:
+		return "http/2"
+	case GRPC:
+		return "http/2"
+	default:
+		return ""
+	}
+}
+
+func matchListener(listeners []*listener.Listener, input Call) *listener.Listener {
+	// First find exact match for the IP/Port, then fallback to wildcard IP/Port
+	// There is no wildcard port
+	for _, l := range listeners {
+		if matchAddress(l.GetAddress(), input.Address, input.Port) {
+			return l
+		}
+	}
+	for _, l := range listeners {
+		if matchAddress(l.GetAddress(), "0.0.0.0", input.Port) {
+			return l
+		}
+	}
+	return nil
```

Comment:
todo: if none of 0.0.0.0 matches, then you have to fallback to the original dst listener where we have passthrough. Otherwise, this logic wont match what we have in listener code today. Also, might be worthwhile distinguishing gateway vs sidecar.

Reply:
Good call. Right now this ignores that one.